### DISCUSSION TITLE
how to set context in machine with context input and assign actions

### DISCUSSION THREAD

**Discussion by codemonkeynorth** [statelyai/xstate#4795] [Q&A] [answered]
Hi,

hi, I want to pass context to my machine when i create it like this inside my hook:
```ts
import bootMachine from "../fsm/bootMachine"
...
...
const { shouldUseAPI } = useAPI()
const service = createActor(bootMachine, {
  input: {
   shouldUseAPI  // ie dynamic from hook so i need to pass it into my machine 
  },
})
```

and my machine has this which I'm using therefore for the initial context based on that input:
```ts
context: ({ input }) => ({ status: LoadingState.NONE, shouldUseAPI: input?.shouldUseAPI }),
```
but as soon as I run this on eg an `invoke => onError`
```ts
invoke: {
        src: "initialiseAPI",
        onDone: {
          target: "initialise_user",
        },
        onError: {
          actions: assign({ status: LoadingState.ERROR }),
          target: "initialise_api_error",
        },
      },
```

then my context's  `shouldUseAPI` gets set as `undefined `, presumably because it's trying to take `input` which isn't there?

when I use a normal context without input it's fine, but then I can't dynamically set the initial value .

so how do I combine `assign` actions with a `context: ({ input })` ?

thanks for any suggestions
J

**Comment by Andarist** [comment]
If u could create a runnable repro of ur problem (like a codesandbox or stackblitz) then we could quickly fix your issue there.

**Reply by codemonkeynorth** [reply, marked as answer]
sorry I was using both `createActor` (to get a service) but also `useMachine` when I should only really have one of them

thus causing a bunch of conflicts

this works fine:
https://codesandbox.io/p/devbox/great-sid-9n56pd

sorry about that!
